kafka.go

```diff
@@ -254,87 +262,91 @@ func (client *KafkaClient) Register(subscribeBuilder *SubscribeBuilder) error {
-			log.Errorf("unable to register callback. error: %v", err)
-			return
-		}
+	isRegistered, err := client.registerSubscriber(subscribeBuilder)
+	if err != nil {
+		log.Errorf("unable to register callback. error: %v", err)
+		return err
+	}
 
-		if isRegistered {
-			log.Warnf("topic and event already registered. topic: %s , event: %s", subscribeBuilder.topic,
-				subscribeBuilder.eventName)
-		}
+	if isRegistered {
+		return fmt.Errorf(
+			"topic and event already registered. topic: %s , event: %s",
+			subscribeBuilder.topic,
+			subscribeBuilder.eventName,
+		)
+	}
 
-		config := client.subscribeConfig
-		config.Topic = topic
-		config.GroupID = groupID
-		config.StartOffset = kafka.LastOffset
-		config.MaxWait = kafkaMaxWait
-		reader := kafka.NewReader(config)
+	config := client.subscribeConfig
+	config.Topic = topic
+	config.GroupID = groupID
+	config.StartOffset = kafka.LastOffset
+	config.MaxWait = kafkaMaxWait
+	reader := kafka.NewReader(config)
 
-		defer func() {
-			_ = reader.Close()
-		}()
+	go func() {
+		defer reader.Close() // nolint: errcheck
+		defer client.unregister(subscribeBuilder)
 
 		for {
-			consumerMessage, err := reader.ReadMessage(subscribeBuilder.ctx)
-			if err != nil {
-				log.Error("unable to subscribe topic from kafka. error: ", err)
+			select {
+			case <-subscribeBuilder.ctx.Done():
+				subscribeBuilder.callback(subscribeBuilder.ctx, nil, subscribeBuilder.ctx.Err())
 				return
-			}
+			default:
+				consumerMessage, errRead := reader.ReadMessage(subscribeBuilder.ctx)
+				if errRead != nil {
+					log.Error("unable to subscribe topic from kafka. error: ", errRead)
+					return
+				}
 
-			go client.processMessage(consumerMessage, groupID)
+				client.processMessage(subscribeBuilder, consumerMessage)
+			}
 		}
 	}()
 
 	return nil
 }
 
-// registerCallback add callback to map with topic and eventName as a key
-func (client *KafkaClient) registerCallback(topic, eventName string, callback func(event *Event, err error)) (
-	isRegistered bool, err error) {
-	if innerMap, ok := client.subscribeMap.Load(topic); ok {
-		innerValue, ok := innerMap.(*sync.Map)
-		if !ok {
-			return false, errors.New("unable to convert interface to sync map")
-		}
-
-		if _, ok = innerValue.Load(eventName); ok {
-			return true, nil
-		}
-
-		// event callback with this topic name is there with another event callback registered, so append it
-		if innerValue != nil {
-			innerValue.Store(eventName, callback)
-			client.subscribeMap.Store(topic, innerValue)
-
-			return false, nil
+// registerSubscriber add callback to map with topic and eventName as a key
+func (client *KafkaClient) registerSubscriber(subscribeBuilder *SubscribeBuilder) (
+	isRegistered bool,
+	err error,
+) {
+	client.lock.Lock()
+	defer client.lock.Unlock()
+
+	for subs := range client.subscribers {
+		if subs.topic == subscribeBuilder.topic && subs.eventName == subscribeBuilder.eventName {
+			if subscribeBuilder.groupID == "" {
+				return true, nil
+			}
```

Comment:
what happens if the groupID isn't empty?

Reply:
@fadhillahentino 
if the groupID is empty:
 - if we already have a subscription on the topic+eventName -> error(subscription already exists)
 - if we don't have a subscription on the topic+eventName -> subscription created(without group id)

if the groupID is not empty:
 - subscription will be crated into the groupID

Reply:
Let say we have groupID A and then we registered groupID A again with different callback so we have 2 object that represent groupID A, is it intended?

Reply:
@fadhillahentino if we are talking about subscription-groups - in case we we have multiple subscribers(consumers) to the same subscription group(workers pool pattern) -  the message would be delivered to a random callback function(random worker) from the same subscription-group:
- https://kafka.apache.org/intro#intro_consumers
- https://blog.cloudera.com/scalability-of-kafka-messaging-using-consumer-groups/

Reply:
yes I know that, so it will intended if we have 2 different callback with same groupID, right?

Reply:
if we have multiple consumers on the same `topic + event-name + group-id` the message will be delivered to only one randomly selected callback

Reply:
okay
another question, I think we might have multiple goroutine that subscribe in the same topic, am I right?.

Reply:
@fadhillahentino only if these goroutines subscriber to a not-empty group-id. Without group-id we don't allow to subscribe second callback(previous behavior)

Reply:
I see, but if we have a topic with different event name, will it create many goroutines?